### PR TITLE
deprecate vectorized methods hiding in Dates

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request gleefully deprecates a few vectorized methods that were hiding in `Dates`. Addresses #23167. Best!

### CODE REVIEW

**Review by ararslan — APPROVED** [review]

File: test/dates/io.jl [hunk 1]
```diff
@@ -323,15 +323,14 @@ f = "ymd"
 dr = ["2000-01-01", "2000-01-02", "2000-01-03", "2000-01-04", "2000-01-05",
       "2000-01-06", "2000-01-07", "2000-01-08", "2000-01-09", "2000-01-10"]
 dr2 = [Dates.Date(2000) : Dates.Date(2000, 1, 10);]
-@test Dates.Date(dr) == dr2
-@test Dates.Date(dr, "yyyy-mm-dd") == dr2
+@test Dates.Date.(dr) == dr2
+@test Dates.Date.(dr, "yyyy-mm-dd") == dr2
```
Comment (omus) [inline test/dates/io.jl]: Tests will be more performant if we change this to use the `dateformat` string macro.

Comment (Sacha0) [inline test/dates/io.jl]: Changed these `Date` and `DateTime` tests to use the `dateformat` string macro. Thanks! :)

**Comment by omus** [comment]
I forgot to post this last night. The broadcasting performance when you are not using the `dateformat` string macro is pretty terrible:

```julia
julia> using BenchmarkTools

julia>  dr = ["2000-01-01", "2000-01-02", "2000-01-03", "2000-01-04", "2000-01-05",
              "2000-01-06", "2000-01-07", "2000-01-08", "2000-01-09", "2000-01-10"];

julia> @benchmark Date($dr, "yyyy-mm-dd")  # deprecated
BenchmarkTools.Trial:
  memory estimate:  4.47 KiB
  allocs estimate:  104
  --------------
  minimum time:     17.368 μs (0.00% GC)
  median time:      19.149 μs (0.00% GC)
  mean time:        23.571 μs (16.24% GC)
  maximum time:     32.019 ms (94.77% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark Date($dr, dateformat"yyyy-mm-dd")  # deprecated
BenchmarkTools.Trial:
  memory estimate:  528 bytes
  allocs estimate:  13
  --------------
  minimum time:     465.571 ns (0.00% GC)
  median time:      473.120 ns (0.00% GC)
  mean time:        567.269 ns (6.44% GC)
  maximum time:     141.774 μs (99.63% GC)
  --------------
  samples:          10000
  evals/sample:     196

julia> @benchmark Date.($dr, "yyyy-mm-dd")
BenchmarkTools.Trial:
  memory estimate:  40.86 KiB
  allocs estimate:  949
  --------------
  minimum time:     182.903 μs (0.00% GC)
  median time:      191.881 μs (0.00% GC)
  mean time:        205.410 μs (4.52% GC)
  maximum time:     32.989 ms (93.24% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark Date.($dr, dateformat"yyyy-mm-dd")
BenchmarkTools.Trial:
  memory estimate:  480 bytes
  allocs estimate:  11
  --------------
  minimum time:     443.898 ns (0.00% GC)
  median time:      449.193 ns (0.00% GC)
  mean time:        488.670 ns (7.16% GC)
  maximum time:     143.805 μs (99.62% GC)
  --------------
  samples:          10000
  evals/sample:     197
```
Maybe we should deprecate the `Date(::AbstractString, ::AbstractString)` constructor as well. That can be for another PR.

**Comment by quinnj** [comment]
I agree @omus, we should deprecate the string method. This looks great, thanks.

**Comment by Sacha0** [comment]
Thanks all! :)

File: test/dates/io.jl [hunk 2]
```diff
@@ -319,19 +319,18 @@ f = "ymd"
...
 dr2 = [Dates.Date(2000) : Dates.Date(2000, 1, 10);]
-@test Dates.Date(dr) == dr2
-@test Dates.Date(dr, "yyyy-mm-dd") == dr2
+@test Dates.Date.(dr) == dr2
+@test Dates.Date.(dr, dateformat"yyyy-mm-dd") == dr2
 @test Dates.DateTime.(dr) == Dates.DateTime.(dr2)
-@test Dates.DateTime(dr, "yyyy-mm-dd") == Dates.DateTime.(dr2)
+@test Dates.DateTime.(dr, dateformat"yyyy-mm-dd") == Dates.DateTime.(dr2)
 
-@test Dates.format(dr2) == dr
```
Comment (tkelman) [inline test/dates/io.jl]: could this test have been kept but with a dot, or is there no 1-arg `format` any more?

**Pull request merged**